fraud/service.go

```diff
@@ -110,35 +138,182 @@ func (f *service) processIncoming(
 		f.pubsub.BlacklistPeer(from)
 		return pubsub.ValidationReject
 	}
-	log.Warnw("received fraud proof", "proofType", proof.Type(),
-		"height", proof.Height(),
-		"hash", hex.EncodeToString(extHeader.DAH.Hash()),
-		"from", from.String(),
-	)
-	msg.ValidatorData = proof
+
+	err = f.put(ctx, proof.Type(), hex.EncodeToString(proof.HeaderHash()), msg.Data)
+	if err != nil {
+		log.Error("error while adding a fraud proof to storage: ", err)
+	}
+
+	return pubsub.ValidationAccept
+}
+
+func (f *ProofService) Get(ctx context.Context, proofType ProofType) ([]Proof, error) {
 	f.storesLk.Lock()
 	store, ok := f.stores[proofType]
 	if !ok {
 		store = initStore(proofType, f.ds)
 		f.stores[proofType] = store
 	}
 	f.storesLk.Unlock()
-	err = put(ctx, store, string(proof.HeaderHash()), msg.Data)
-	if err != nil {
-		log.Error(err)
-	}
-	log.Warn("Shutting down services...")
-	return pubsub.ValidationAccept
+
+	return getAll(ctx, store, proofType)
 }
 
-func (f *service) Get(ctx context.Context, proofType ProofType) ([]Proof, error) {
+func (f *ProofService) put(ctx context.Context, proofType ProofType, hash string, data []byte) error {
 	f.storesLk.Lock()
 	store, ok := f.stores[proofType]
 	if !ok {
 		store = initStore(proofType, f.ds)
 		f.stores[proofType] = store
 	}
 	f.storesLk.Unlock()
+	return put(ctx, store, hash, data)
+}
 
-	return getAll(ctx, store, proofType)
+func (f *ProofService) handleFraudMessageRequest(stream network.Stream) {
```

Comment:
Stream is not closed. Let's follow the pattern from another comment. But here we read first so we `CloseRead` first and then `CloseWrite`